### PR TITLE
Feature/#73 entity kind

### PR DESCRIPTION
**Pull request by pbimpisidis**

### CODE REVIEW

File: it/EntityKind.Tests.ps1 [hunk 1]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
 		
-		It "CreateEntityKindWithoutName-Fail" -Test {
+		It "CreateEntityKindWithoutName-ShouldFail" -Test {
 			try
 			{
-				# Arrange
-				$entityKindName = ("TestForPesterRun-{0}" -f [Guid]::NewGuid().ToString())
-				$entityKindDescription = ("Test Description For Pester Run - {0}" -f [Guid]::NewGuid().ToString())
+				#ARRANGE
+				$entityKindName = $null;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis remove this variable

File: it/EntityKind.Tests.ps1 [hunk 2]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$entityKindDescription = ("Test Description For Pester Run - {0}" -f [Guid]::NewGuid().ToString());
 				$entityKindVersion = [Guid]::NewGuid().ToString();
-				$entityKindParameters = "'{ }'"
+				$entityKindParameters = "'{ }'";
+				
+				$newEntityKind = New-Object biz.dfch.CS.Appclusive.Api.Core.EntityKind;
+				$newEntityKind.Name = $EntityKindName;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis I suggest removing this line as it does nothing. When creating a new object the `Name` is by default set to `$null`

File: it/EntityKind.Tests.ps1 [hunk 3]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$newEntityKind.Description = $EntityKindDescription;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 4]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$newEntityKind.Version = $EntityKindVersion;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 5]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$newEntityKind.Parameters = $EntityKindParameters;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 6]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
 			try
 			{
-				# Arrange
-				$entityKindName = ("TestForPesterRun-{0}" -f [Guid]::NewGuid().ToString())
+				#ARRANGE
+				$entityKindName = $entityPrefix + "EntityKind";
 				$entityKindDescription = ("Test Description For Pester Run - {0}" -f [Guid]::NewGuid().ToString())
-				$entityKindVersion = [Guid]::NewGuid().ToString();
-				$entityKindParameters = "'{ }'"
+				$entityKindVersion = $null;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis remove this variable

File: it/EntityKind.Tests.ps1 [hunk 7]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$entityKindParameters = "'{ }'";
 				
-				# Act
-				$entityKind = CreateEntityKind -entityKindName $entityKindName -entityKindDescription $entityKindDescription -entityKindVersion $entityKindVersion
-				$entityKind.Version = $null;
-				$svc.Core.AddToEntityKinds($entityKind)
+				$newEntityKind = New-Object biz.dfch.CS.Appclusive.Api.Core.EntityKind;
+				$newEntityKind.Name = $EntityKindName;
+				$newEntityKind.Description = $EntityKindDescription;
+				$newEntityKind.Version = $EntityKindVersion;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis remove this line

File: it/EntityKind.Tests.ps1 [hunk 8]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
-				$entityKindParameters = "'{ }'"
+				$entityKindVersion = $null;
+				$entityKindParameters = "'{ }'";
 				
-				# Act
-				$entityKind = CreateEntityKind -entityKindName $entityKindName -entityKindDescription $entityKindDescription -entityKindVersion $entityKindVersion
-				$entityKind.Version = $null;
-				$svc.Core.AddToEntityKinds($entityKind)
+				$newEntityKind = New-Object biz.dfch.CS.Appclusive.Api.Core.EntityKind;
+				$newEntityKind.Name = $EntityKindName;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 9]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$newEntityKind.Description = $EntityKindDescription;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 10]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$newEntityKind.Version = $EntityKindVersion;
+				$newEntityKind.Parameters = $EntityKindParameters;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis local variables should start with lowercase

File: it/EntityKind.Tests.ps1 [hunk 11]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
-					$svc.Core.DeleteObject($entityKind);
-					$result = $svc.Core.SaveChanges();
-					$result.StatusCode | Should Be 204;
-					write-warning "EntityKind deleted"
-				}
 			}
 		}
 		
-		It "CreateEntityKindTwiceWithDiffrentVersion" -Test {
+		It "CreateEntityKindTwiceWithDifferentVersion" -Test {
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis I suggest renaming as follows `CreateAnotherVersionOfTheSameEntityKind-ShouldSucceed`

File: it/EntityKind.Tests.ps1 [hunk 12]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
-				#Cleanup
-				$svc.Core.DeleteObject($entityKind2);
-				$result = $svc.Core.SaveChanges();
-				$result.StatusCode | Should Be 204;
-				write-warning "EntityKind deleted"
 			}
 		}
 		
-		It "CreateEntityKindTwiceWithSameNameAndVersion-Fail" -Test {
+		It "CreateEntityKindsWithSameNameAndVersion-ShouldFail" -Test {
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis I suggest renaming this test as follows: `CreateEntityKindWithSameNameAndVersionTwice-ShouldFail`

File: it/EntityKind.Tests.ps1 [hunk 13]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$entityKind2.Version = $EntityKindVersion;
+		
+				#ACT add it to entity kinds - should throw
 				$svc.Core.AddToEntityKinds($entityKind2);
-				
-				
-				{ $svc.Core.SaveChanges() } | Should Throw;
-				$svc.Core.RevertEntityState($entityKind2)
-				$Error[0].Exception | Should Match ("EntityKind with Name '{0}' and Version '{1}' already exists" -f $entityKindName, $entityKindVersion)
+				{ $result = $svc.Core.SaveChanges(); } | Should Throw;
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis I suggest checking for the DataServiceClientException as you did in the tests before 

`{ $result = $svc.Core.SaveChanges(); } | Should ThrowDataServiceClientException @{StatusCode = 400};`

File: it/EntityKind.Tests.ps1 [hunk 14]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
-				#Cleanup
-				$svc.Core.DeleteObject($entityKind);
-				$result = $svc.Core.SaveChanges();
-				$result.StatusCode | Should Be 204;
-				write-warning "EntityKind deleted"
 			}
 		}
 		
-		It "UpdateEntityKindNameDescriptionVersionParameter-Succeeds" -Test {
+		It "UpdateEntityKindNameDescriptionVersionParameter" -Test {
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis I suggest renaming this test to `UpdateEntityKindSucceeds`

File: it/EntityKind.Tests.ps1 [hunk 15]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
-				$entityKindUpdateDescription = ("DescriptionUpdated-{0}" -f [Guid]::NewGuid().ToString());
-				$entityKindUpdateParameter = ("ParameterUpdated-{0}" -f [Guid]::NewGuid().ToString());
+				# ARRANGE Update
+				$entityKindUpdateName = $entityPrefix + "EntityKind Updated";
+				$entityKindUpdateDescription = $entityKindDescription + "Updated";
 				$entityKindUpdateVersion = 2;
+				$entityKindUpdateParameter = "Updated";
 				
-				#Act
+				#ACT set new parameters
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis do not use property name in comment as you do not only set `Parameters`. I suggest writing `#ACT update properties`

File: it/EntityKind.Tests.ps1 [hunk 16]
```diff
@@ -294,26 +235,17 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
 				$svc.Core.UpdateObject($entityKind);
 				$result = $svc.Core.SaveChanges();
 				
-				# Assert	
-				write-warning $entityKindCreated.Name
-				$entityKindUpdated = $svc.Core.EntityKinds.AddQueryOption('$filter', "Id eq "+$entityKind.Id+"")
-
+				# Assert
+				$entityKindUpdated = $svc.Core.EntityKinds.AddQueryOption('$filter', "Id eq "+$entityKind.Id+"");
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis replace this line with the following code

```posh
$updatedEntityKind = Get-ApcEntityKind -Id $entityKind.Id
```

File: it/EntityKind.ps1 [hunk 17]
```diff
@@ -1,14 +1,52 @@
...
+	$newEntityKind.Version = $EntityKindVersion;
+	$newEntityKind.Parameters = $EntityKindParameters;
+	
+	#add it to entity Kinds
+	$svc.Core.AddToEntityKinds($newEntityKind);
+	$result = $svc.Core.SaveChanges();
+	
+	#get entity Kind
+	$query = "Id eq {0}" -f $newEntityKind.Id;
+	$entityKind = $svc.Core.EntityKinds.AddQueryOption('$filter', $query) | Select;
```
Comment (rufer7) [inline it/EntityKind.ps1]: @pbimpisidis use the `Get-ApcEntityKind` Cmdlet

File: it/EntityKind.ps1 [hunk 18]
```diff
@@ -1,14 +1,52 @@
...
+	}
+	$bin = $entityKind.Version | Should Be $EntityKindVersion;
+	if($EntityKindParameters)
+	{
+	$bin = $entityKind.Parameters | Should Be $EntityKindParameters;
+	}
+	
+	$bin = Pop-ApcChangeTracker -Svc $Svc;
+
+	$bin = $svc.Core.AttachIfNeeded($entityKind); #attaches a detached entity to the ChangeTracker if not already attached
```
Comment (rufer7) [inline it/EntityKind.ps1]: @pbimpisidis please put comments on the line above the code and not on the same line...

File: it/EntityKind.Tests.ps1 [hunk 19]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$entityKind2 = Create-EntityKind -entityKindName $entityKindName -entityKindDescription $entityKindDescription2 -entityKindVersion $entityKindVersion2 -svc $svc;
+				$entityKindId2 = $entityKind2.Id;
 				
-				$entityKind1Check = $svc.Core.EntityKinds.AddQueryOption('$filter',("Id eq {0}" -f $entityKind1.Id))
-				$entityKind2Check = $svc.Core.EntityKinds.AddQueryOption('$filter',("Id eq {0}" -f $entityKind2.Id))
-				$entityKind1Check.Description | Should Not Be $entityKind2Check.Description
-				$entityKind1Check.Id | Should Not Be $entityKind2Check.Id
-				$entityKind1Check.Name | Should Be $entityKind2Check.Name
+				#ASSERT
+				$entityKind1Check = $svc.Core.EntityKinds.AddQueryOption('$filter',("Id eq {0}" -f $entityKindId1));
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis use the `Get-ApcEntityKind` Cmdlet (Do not forget passing `-svc`)

File: it/EntityKind.Tests.ps1 [hunk 20]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+				$entityKind2Check = $svc.Core.EntityKinds.AddQueryOption('$filter',("Id eq {0}" -f $entityKindId2));
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis use the `Get-ApcEntityKind` Cmdlet (Do not forget passing `-svc`)

File: it/EntityKind.Tests.ps1 [hunk 21]
```diff
@@ -63,229 +79,154 @@ Describe -Tags "EntityKind.Tests" "EntityKind.Tests" {
...
+		
+				#ACT add it to entity kinds - should throw
 				$svc.Core.AddToEntityKinds($entityKind2);
-				
-				
-				{ $svc.Core.SaveChanges() } | Should Throw;
-				$svc.Core.RevertEntityState($entityKind2)
-				$Error[0].Exception | Should Match ("EntityKind with Name '{0}' and Version '{1}' already exists" -f $entityKindName, $entityKindVersion)
+				{ $result = $svc.Core.SaveChanges(); } | Should Throw;
+				$svc.Core.RevertEntityState($entityKind2) # reverts the state of entity to EntityStates.Detatched
```
Comment (rufer7) [inline it/EntityKind.Tests.ps1]: @pbimpisidis please put comments on the line above the code and not on the same line...

**Comment by pbimpisidis** [comment]
@rufer7 updated code, all green

**Pull request merged**